trieste/models/gpflow/models.py

```diff
@@ -112,34 +114,52 @@ def covariance_between_points(
 
         .. math:: \Sigma_{12} = K_{12} - K_{x1}(K_{xx} + \sigma^2 I)^{-1}K_{x2}
 
-        :param query_points_1: Set of query points with shape [N, D]
+        :param query_points_1: Set of query points with shape [..., N, D]
         :param query_points_2: Sets of query points with shape [M, D]
-
-        :return: Covariance matrix between the sets of query points with shape [N, M]
+        :return: Covariance matrix between the sets of query points with shape [..., L, N, M]
         """
         if isinstance(self.model, SGPR):
             raise NotImplementedError("Covariance between points is not supported for SGPR.")
 
-        tf.debugging.assert_shapes([(query_points_1, ["N", "D"]), (query_points_2, ["M", "D"])])
+        tf.debugging.assert_shapes(
+            [(query_points_1, [..., "N", "D"]), (query_points_2, ["M", "D"])]
+        )
 
         x = self.model.data[0].value()
         num_data = tf.shape(x)[0]
         s = tf.linalg.diag(tf.fill([num_data], self.model.likelihood.variance))
 
-        K = self.model.kernel(x)
-        L = tf.linalg.cholesky(K + s)
+        K = self.model.kernel(x)  # [num_data, num_data] or [L, num_data, num_data]
+        Kx1 = self.model.kernel(query_points_1, x)  # [..., N, num_data] or [..., L, N, num_data]
+        Kx2 = self.model.kernel(x, query_points_2)  # [num_data, M] or [L, num_data, M]
+        K12 = self.model.kernel(query_points_1, query_points_2)  # [..., N, M] or [..., L, N, M]
+
+        if len(tf.shape(K)) == 2:
+            K = tf.expand_dims(K, -3)
+            Kx1 = tf.expand_dims(Kx1, -3)
+            Kx2 = tf.expand_dims(Kx2, -3)
+            K12 = tf.expand_dims(K12, -3)
+        elif len(tf.shape(K)) > 3:
+            raise NotImplementedError(
+                "Covariance between points is not supported "
+                f"for kernels of type "
+                f"{type(self.model.kernel)}."
+            )
 
-        Kx1 = self.model.kernel(x, query_points_1)
-        Linv_Kx1 = tf.linalg.triangular_solve(L, Kx1)
+        L = tf.linalg.cholesky(K + s)  # [L, num_data, num_data]
 
-        Kx2 = self.model.kernel(x, query_points_2)
-        Linv_Kx2 = tf.linalg.triangular_solve(L, Kx2)
+        Kx1 = leading_transpose(Kx1, [..., -1, -2])  # [..., L, num_data, N]
+        Linv_Kx1 = tf.linalg.triangular_solve(L, Kx1)  # [..., L, num_data, N]
+        Linv_Kx2 = tf.linalg.triangular_solve(L, Kx2)  # [L, num_data, M]
 
-        K12 = self.model.kernel(query_points_1, query_points_2)
-        cov = K12 - tf.tensordot(tf.transpose(Linv_Kx1), Linv_Kx2, [[-1], [-2]])
+        cov = K12 - tf.einsum("...lji,ljk->...lik", Linv_Kx1, Linv_Kx2)  # [..., L, N, M]
 
         tf.debugging.assert_shapes(
-            [(query_points_1, ["N", "D"]), (query_points_2, ["M", "D"]), (cov, ["N", "M"])]
+            [
+                (query_points_1, [..., "N", "D"]),
+                (query_points_2, ["M", "D"]),
+                (cov, [..., "L", "N", "M"]),
+            ]
```

Comment:
Since this is the outcome of the calculation that isn't directly in user's control, if we were to throw an exception here, let's at least give a clue in the message as to why.

Reply:
any suggestion about how to do this?

Reply:
I can only answer in very general terms. Why would you expect this assert to ever fail? If this can be formulated in a phrase, that's your error message. At the very least "received unexpected shapes after calculating blah and blah" shall suffice, if we can't say anything more useful.